trestle/core/rules_validator.py

```diff
@@ -0,0 +1,110 @@
+        """
+        for param in as_list(item.set_parameters if not isinstance(item, cat.Control) else item.params):
+            parameter_id = param.param_id if not isinstance(item, cat.Control) else param.id
+            control = cat_int.get_control_by_param_id(parameter_id)
+            if not control:
+                deep_set(shared_params, [parameter_id, comp_name, control_id], param.values)
+
+    def model_is_valid(self, model: TopLevelOscalModel, quiet: bool, trestle_root: pathlib.Path = '') -> bool:
+        """
+        Test if the model is valid.
+
+        args:
+            model: A top level OSCAL model.
+            quiet: Don't report msgs unless invalid.
+            trestle_root: Trestle root path.
+
+        returns:
+            True (valid) if the model's rule parameter values are the same accross controls.
+        """
+        # verify if model type is either an SSP of a Component Definition
+        if not isinstance(model, (SystemSecurityPlan)):
+            return True
+
+        components = as_list(model.system_implementation.components)
+        component_titles = [x.title for x in components]
+        profile_resolver = ProfileResolver()
+        catalog = profile_resolver.get_resolved_profile_catalog(trestle_root, model.import_profile.href)
+        catalog_interface = CatalogInterface(catalog)
+
+        for control in catalog_interface.get_all_controls_from_catalog(True):
+            self.add_shared_param(control, control.id, catalog_interface)
+        # iterate by each implemented requirement defined
+        for imp_req in as_list(model.control_implementation.implemented_requirements):
+            # iterate by each by_component in each implemented requirement
+            for by_component in as_list(imp_req.by_components):
+                component = get_item_from_list(components, by_component.component_uuid, lambda x: x.uuid)
+                self.add_shared_param(imp_req, imp_req.control_id, catalog_interface, component.title)
+                self.add_shared_param(by_component, imp_req.control_id, catalog_interface, component.title)
+            # includes rule param values in each by_component present in statements
+            for statement in as_list(imp_req.statements):
+                for by_comp in as_list(statement.by_components):
+                    component = get_item_from_list(components, by_component.component_uuid, lambda x: x.uuid)
+                    self.add_shared_param(by_comp, imp_req.control_id, catalog_interface, component.title)
+        if not shared_params:
+            return True
+        # compare all values in shared paramerets by component basis
+        for shared_param, _ in as_list(shared_params.items()):
+            for comp_title in component_titles:
+                # retrieve all control rule param values attached to a component
+                controls_list = deep_get(shared_params, [shared_param, comp_title], None)
+                if not controls_list:
```

Comment:
This overall logic can be simplified.  The dict is a dict of `{paramid: {comp_id: {control_id: values}}}`  so just grab all its values of the form `{comp_id: {control_id: values}}` and the values for each control id should be the same.

Reply:
done resolved